### PR TITLE
dfp: Add DFP recipe.

### PR DESCRIPTION
**Pull request by agdavydov81**
Add initial version if the Conan package for the [DFP](https://github.com/epam/DFP/)/1.0.2

C/C++ support is a new feature of the DFP (Decimal Floating Point) public library.
This library is developed by the [EPAM](https://github.com/epam/) company under Apache-2.0 and Intel licenses.

---

- [+] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [+] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [+] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/dfp/all/native/test_package/conanfile.py
Comment (RubenRBS) [inline recipes/dfp/all/native/test_package/conanfile.py:1]: CCI expects the tests to be under (usually), a `recipes/dfp/all/test_package` folder - if you need to differenciate between extra test packages, the current recommended approach is to name the folder something like `recipes/dfp/all/test_native_package`

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Thanks a lot! Quick feedback on the test name so the bot stops complaining :)

File: recipes/dfp/all/native/test_package/conanfile.py
Comment (agdavydov81) [reply recipes/dfp/all/native/test_package/conanfile.py:1]: Thank you, I will try move tests.

Comment (agdavydov81) [reply recipes/dfp/all/native/test_package/conanfile.py:1]: @RubenRBS Hi!
Hi have moved all sources to the top directory, so, all must be fine, but some checks still fail:
c3i/conan-v2/pr-merge — This commit cannot be built
continuous-integration/jenkins/pr-merge — This commit cannot be built

Also I can't get the error message.
Could you please copy error messages to me?

Comment (agdavydov81) [reply recipes/dfp/all/native/test_package/conanfile.py:1]: Bot doesn't post any messages.

**Comment by franramirez688** [comment]
Hi @agdavydov81 

Thanks a lot for contributing! I just pushed a couple of changes to help you and move it forward.

* Added source method to download the source code in the Conan cache.
* Removed `test_v1_package` as it's not necessary anymore for new Conan recipes.

Let us know if you need more help. We'll review it after the CI finishes

**Comment by agdavydov81** [comment]
@franramirez688  Thank you!
I will try to downgrade my cmake_minimum_required to 3.15

**Comment by franramirez688** [comment]
> @franramirez688 Thank you! I will try to downgrade my cmake_minimum_required to 3.15

Oh, if you're saying that because of the minimum required version, don't worry about it because we can force the CMake version used, putting it as a tool requirement. Let me show you how.

**Comment by agdavydov81** [comment]
@franramirez688 Thank you for your fixes. I'm was busy today on another project...

**Comment by agdavydov81** [comment]
@franramirez688 @RubenRBS 
Greet news! All checks are passed!
Thank you very much for your help!

**Review by franramirez688 — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]
LGTM!

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Blocking until I have the time to review :)

**Comment by agdavydov81** [comment]
@RubenRBS Hi!
Could you please review PR on this week?

File: recipes/dfp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    # Optional metadata
+    description = "Decimal Floating Point Arithmetic Library"
+    homepage = "https://github.com/epam/DFP/"
+    url = "https://github.com/conan-io/conan-center-index"
+    license = ("Apache-2.0", "Intel")
+    topics = ("decimal", "dfp", "ieee-754", "deltix")
+
+    # Binary configuration
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
```
Comment (RubenRBS) [inline recipes/dfp/all/conanfile.py:23]: Conan recipes usually allow to modify position independent code via the `fPIC` option. I see in your upstream CMakeLists that [it's always turned ON](https://github.com/epam/DFP/blob/master/native/CMakeLists.txt#L76), but unless not supported by the library, we usually prefer to give the end user the option to enable-disable it by themsevlves.

Let me know if it's ok to add the option and I'll do that myself :)

File: recipes/dfp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    default_options = {"shared": False}
+
+    # Sources are located in the same place as this recipe, copy them to the recipe
+    exports_sources = "CMakeLists.txt", "src/*", "include/*"
```
Comment (RubenRBS) [inline recipes/dfp/all/conanfile.py:27]: This line is doing nothing - those files are not in this folder

File: recipes/dfp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    def source(self):
```
Comment (RubenRBS) [inline recipes/dfp/all/conanfile.py:29]: In the upstream CMakeLists.txt, [you've set C++ 11 as the minimum required version ](https://github.com/epam/DFP/blob/master/native/CMakeLists.txt#L77) - You'll want to add [something like this check in the `validate()` method](https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py#L82) to ensure C++11 is used

**Review by RubenRBS — COMMENTED** [review]
Quick question about fPIC support and a few suggested changes, thanks a lot for your contribution :)

File: recipes/dfp/all/conanfile.py (hunk 3 above)
Comment (agdavydov81) [reply recipes/dfp/all/conanfile.py:33]: You right - this is just part of one for old version.
This library do not requires C++ at all.
I have pushed fix.

File: recipes/dfp/all/conanfile.py (hunk 2 above)
Comment (agdavydov81) [reply recipes/dfp/all/conanfile.py]: OK, I'll delete this line.

**Comment by agdavydov81** [comment]
@RubenRBS I have fixed your notes.
Please review again.

**Review by franramirez688 — APPROVED** [review]

**Review by RubenRBS — APPROVED** [review]

**Comment by agdavydov81** [comment]
@RubenRBS  @franramirez688 
Thank you so much for your help!!!

**Pull request merged**